server/src/weaverbird/pipeline/steps/utils/combination.py

```diff
@@ -95,20 +95,35 @@ async def resolve_if_reference(
     reference_resolver: ReferenceResolver,
     pipeline_or_domain_name_or_ref: PipelineOrDomainNameOrReference,
 ) -> PipelineOrDomainName | None:
-    from weaverbird.pipeline.pipeline import PipelineWithRefs, ReferenceUnresolved
+    from weaverbird.pipeline.pipeline import ReferenceUnresolved
 
     if isinstance(pipeline_or_domain_name_or_ref, Reference):
-        pipeline_or_domain_name = await reference_resolver(pipeline_or_domain_name_or_ref)
-        if isinstance(pipeline_or_domain_name, list):
-            # Recursively resolve any reference in sub-pipelines
-            pipeline = PipelineWithRefs(steps=pipeline_or_domain_name)
-
-            try:
-                pipeline_without_refs = await pipeline.resolve_references(reference_resolver)
-                return pipeline_without_refs.model_dump()["steps"]
-            except ReferenceUnresolved:
-                return None  # skip
-        else:
-            return pipeline_or_domain_name
+        try:
+            pipeline_or_domain_name = await reference_resolver(pipeline_or_domain_name_or_ref)
+            if isinstance(pipeline_or_domain_name, list):
+                return await _resolve_references_in_pipeline(reference_resolver, pipeline_or_domain_name)
+            else:
+                return pipeline_or_domain_name
+        except ReferenceUnresolved:
+            return None  # skip
+
+    if isinstance(pipeline_or_domain_name_or_ref, list):
+        return await _resolve_references_in_pipeline(reference_resolver, pipeline_or_domain_name_or_ref)
     else:
         return pipeline_or_domain_name_or_ref
+
+
+async def _resolve_references_in_pipeline(
+    reference_resolver: ReferenceResolver,
+    pipeline: list,
```

Comment:
Could you please improve the typing ? A list of what ? Given the condition l.122, can it be something else  ?

Reply:
[chore: code review feedback](https://github.com/ToucanToco/weaverbird/pull/1995/commits/3696cdac5d823b1b4c826cc7fe17369e51566380)